framework/src/service/ServiceReferenceBasePrivate.cpp

```diff
@@ -115,52 +126,42 @@ InterfaceMapConstPtr ServiceReferenceBasePrivate::GetServiceInterfaceMap(BundleP
 {
   InterfaceMapConstPtr s;
   if (!registration->available) return s;
-
   std::shared_ptr<ServiceFactory> serviceFactory;
-  int count = 0;
 
+  auto l = registration->Lock(); US_UNUSED(l);
```

Comment:
I agree, something needs to call `GetService` on the same service registration. But since a service factory can call into arbitrary framework API, e.g. starting bundles, which then can again do arbitrary things in their activator etc., I think we should make that bullet proof.

Discarding a second instance was the strategy used before. This is an exceptional case and not likely to occur. I would favor no deadlocks instead of optimizing service creation for corner cases :)

To avoid the race you found, would it be enough to check (in the first registration lock scope) if we have a service factory and the usage count is > 0, then return the cached instance? If not, go out of scope and execute the original code.